### ISSUE TITLE
Blackbox tests failing in Jenkins

### ISSUE THREAD

**Issue by michaelestrin** [edgexfoundry/blackbox-testing#282]
From @tsconn23:

> Hi all – This is a heads up that I noticed just now that the blackbox tests in Jenkins stopped working on 18-October.
>
>https://jenkins.edgexfoundry.org/view/blackbox-testing/job/blackbox-testing-centos7-blackbox-4c-2g-master/
>
>If you look at the Console Output for one of the failed reports, you’ll see errors like the following
>
>08:00:43 → Test whether the virtual device service exists
>08:00:43   GET http://edgex-device-virtual:49990/api/v1/ping [errored]
>08:00:43      connect ECONNREFUSED 172.17.0.6:49990
>08:00:43 → Test whether the Random-Boolean-Device device profile has been added to metadata
>08:00:43   GET http://edgex-core-metadata:48081/api/v1/deviceprofile/name/Random-Boolean-Device [errored]
>08:00:43      getaddrinfo ENOTFOUND edgex-core-metadata edgex-core-metadata:48081
>08:01:05      getaddrinfo ENOTFOUND edgex-core-data edgex-core-data:48080
>08:01:05 [errored]
> 
>I went through the Jenkins job definition and then through the relevant scripts in the blackbox-testing repo to see if I could find a commit between the 17th and 18th that would be a candidate for causing the failure, but nothing is jumping out at me.
>
>It’s also my understanding that the blackbox tests are using the docker-compose file below and not the one from developer-scripts/releases/nightly-build/compose-files which has seen a lot of modification.
>
>https://github.com/edgexfoundry/blackbox-testing/blob/master/docker-compose.yml

**Comment by michaelestrin** [comment]
PR did not address issue.  Still seeing numerous failures like `connect ECONNREFUSED 172.17.0.5:49990` and `getaddrinfo ENOTFOUND edgex-core-metadata edgex-core-metadata:48081`.

**Comment by michaelestrin** [comment]
When attempting to deploy to my local Windows machine, several containers appear to have issues:

```
Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$ docker ps -a
CONTAINER ID        IMAGE                                                                  COMMAND                  CREATED             STATUS                     PORTS                                                                                              NAMES
824b3a6affe5        nexus3.edgexfoundry.org:10004/docker-support-rulesengine:1.0.0         "/bin/sh -c 'java -j…"   2 minutes ago       Up 2 minutes               0.0.0.0:48075->48075/tcp                                                                           edgex-support-rulesengine
9b6d8ef5030a        nexus3.edgexfoundry.org:10004/docker-export-distro-go:1.1.0            "/export-distro --re…"   2 minutes ago       Up 2 minutes               0.0.0.0:5566->5566/tcp, 0.0.0.0:48070->48070/tcp                                                   edgex-export-distro
ebbac106433e        nexus3.edgexfoundry.org:10004/docker-device-virtual-go:1.1.0           "/device-virtual --p…"   2 minutes ago       Up 2 minutes               0.0.0.0:49990->49990/tcp                                                                           edgex-device-virtual
169f209593c9        nexus3.edgexfoundry.org:10004/docker-support-scheduler-go:1.1.0        "/support-scheduler …"   2 minutes ago       Exited (1) 2 minutes ago                                                                                                      edgex-support-scheduler
f17350ba819b        nexus3.edgexfoundry.org:10004/docker-core-command-go:1.1.0             "/core-command --reg…"   2 minutes ago       Up 2 minutes               0.0.0.0:48082->48082/tcp                                                                           edgex-core-command
e6daa85e64d4        nexus3.edgexfoundry.org:10004/docker-export-client-go:1.1.0            "/export-client --re…"   2 minutes ago       Up 2 minutes               0.0.0.0:48071->48071/tcp                                                                           edgex-export-client
79dbe4bd9700        nexus3.edgexfoundry.org:10004/docker-core-data-go:1.1.0                "/core-data --regist…"   2 minutes ago       Up 2 minutes               0.0.0.0:5563->5563/tcp, 0.0.0.0:48080->48080/tcp                                                   edgex-core-data
2bb8a1f0f38c        nexus3.edgexfoundry.org:10004/docker-support-notifications-go:1.1.0    "/support-notificati…"   2 minutes ago       Up 2 minutes               0.0.0.0:48060->48060/tcp                                                                           edgex-support-notifications
2b630eda59a5        nexus3.edgexfoundry.org:10004/docker-core-metadata-go:1.1.0            "/core-metadata --re…"   2 minutes ago       Up 2 minutes               0.0.0.0:48081->48081/tcp, 48082/tcp                                                                edgex-core-metadata
09cf80b08d9f        nexus3.edgexfoundry.org:10004/docker-support-logging-go:1.1.0          "/support-logging --…"   2 minutes ago       Up 2 minutes               0.0.0.0:48061->48061/tcp                                                                           edgex-support-logging
08416d2db143        nexus3.edgexfoundry.org:10004/docker-edgex-mongo:1.1.0                 "docker-entrypoint.s…"   2 minutes ago       Up 2 minutes               0.0.0.0:27017->27017/tcp                                                                           edgex-mongo
3ceef869767b        nexus3.edgexfoundry.org:10004/docker-edgex-proxy-go:0.2.1              "./edgexproxy --init…"   2 minutes ago       Exited (0) 2 minutes ago                                                                                                      edgex-proxy
ab77e2f8f78b        nexus3.edgexfoundry.org:10004/docker-edgex-vault-worker-go:0.3.0       "./edgex-vault-worke…"   2 minutes ago       Exited (0) 2 minutes ago                                                                                                      edgex-vault-worker
1c6129dd3c2e        nexus3.edgexfoundry.org:10004/docker-core-config-seed-go:1.1.0         "/edgex/cmd/config-s…"   2 minutes ago       Exited (0) 2 minutes ago                                                                                                      edgex-config-seed
b72998a4b367        nexus3.edgexfoundry.org:10004/docker-edgex-vault:0.3.0                 "docker-entrypoint.s…"   2 minutes ago       Exited (1) 2 minutes ago                                                                                                      edgex-vault
aa2928a3639c        consul:1.3.1                                                           "docker-entrypoint.s…"   2 minutes ago       Up 2 minutes               0.0.0.0:8400->8400/tcp, 8300-8302/tcp, 8301-8302/udp, 8600/tcp, 8600/udp, 0.0.0.0:8500->8500/tcp   edgex-core-consul
5d5be94e1a8f        kong:0.13.0                                                            "/docker-entrypoint.…"   2 minutes ago       Exited (1) 2 minutes ago                                                                                                      kong
56826bdbcc99        nexus3.edgexfoundry.org:10004/docker-edgex-volume:1.0.0                "/bin/sh -c '/usr/bi…"   2 minutes ago       Up 2 minutes                                                                                                                  edgex-files
16c017e4f853        postgres:9.5                                                           "docker-entrypoint.s…"   2 minutes ago       Up 2 minutes               0.0.0.0:5432->5432/tcp                                                                             kong-db
46050662c65e        kong:0.13.0                                                            "/docker-entrypoint.…"   2 minutes ago       Exited (1) 2 minutes ago                                                                                                      kong-migration
455189a8a3d8        nexus3.edgexfoundry.org:10004/docker-app-service-configurable:latest   "/app-service-config…"   2 minutes ago       Up 2 minutes               0.0.0.0:48095->48095/tcp                                                                           edgex-app-service-configurable
9cf129a0484d        postman/newman                                                         "newman"                 2 minutes ago       Exited (0) 2 minutes ago                                                                                                      blackbox-testing_postman_1

Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$ docker logs 169f209593c9
level=ERROR ts=2019-10-21T17:15:18.1891946Z app=edgex-support-scheduler source=main.go:80 msg="edgex-support-scheduler: Service bootstrap failed!"

Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$ docker logs 3ceef869767b
ERROR: 2019/10/21 17:15:17 The status of reverse proxy is unknown with error Get http://kong:8001/: dial tcp 23.202.231.166:8001: connect: connection refused, the initialization is terminated.

Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$ docker logs ab77e2f8f78b
flag provided but not defined: -vaultInterval

Usage: ./edgex-vault-worker [options]
Server Options:
        --consul=true/false                             Indicates if retrieving config from Consul
        --insureskipverify=true/false                   Indicates if skipping the server side SSL cert verifcation, similar to -k of curl
        --init=true/false                               Indicates if security service should be initialized
        --configfile=<file.toml>                        Use a different config file (default: res/configuration.toml)
        --wait=<time in seconds>                Indicates how long the program will pause between the vault initialization until it succeeds
        Common Options:
        -h, --help                                      Show this message


Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$ docker logs 1c6129dd3c2e
level=INFO ts=2019-10-21T17:15:12.5313725Z app=edgex-config-seed source=main.go:54 msg="Service dependencies resolved..."

Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$ docker logs b72998a4b367
/bin/sh: ./security-secrets-setup: not found
Error loading configuration from /vault/config: error loading "/vault/config/local.json": error parsing 'listener': 1 error occurred:

* listeners.tcp: invalid key "tls_perfer_server_cipher_suites" on line 9

Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$ docker logs 5d5be94e1a8f
prefix directory /usr/local/kong not found, trying to create it
2019/10/21 17:15:10 [error] 1#0: init_by_lua error: /usr/local/share/lua/5.1/kong/init.lua:169: connection refused
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/kong/init.lua:169: in function 'init'
        init_by_lua:3: in main chunk
nginx: [error] init_by_lua error: /usr/local/share/lua/5.1/kong/init.lua:169: connection refused
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/kong/init.lua:169: in function 'init'
        init_by_lua:3: in main chunk

Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$ docker logs 46050662c65e
No such command for migrations: bootstrap

Usage: kong migrations COMMAND [OPTIONS]

Manage Kong's database migrations.

The available commands are:
 list   List migrations currently executed.
 up     Execute all missing migrations up to the latest available.
 reset  Reset the configured database (irreversible).

Options:
 -c,--conf        (optional string) configuration file
 -y,--yes         assume "yes" to prompts and run non-interactively
 --v              verbose
 --vv             debug

Michael@DESKTOP-UM6HQHJ MINGW64 ~/source/dell/projects/blackbox-testing/src/github.com/edgexfoundry/blackbox-testing (issue-282)
$

```

**Comment by tsconn23** [comment]
@tingyuz for visibility.

**Comment by michaelestrin** [comment]
Long conversation with @tingyuz, @anonymouse64, @tsconn23, and @jamesrgregg.

Some additional small changes required in a PR for this issue (to follow).

Correcting non-security version of the blackbox tests will require the EDGEX_SECURITY_SECRET_STORE=false environment variable to be set in the non-security version of the Jenkins build.  Related https://github.com/edgexfoundry/ci-management/pull/500.

Correcting security version of the blackbox tests will require a build pipeline to be created for docker-edgex-consul (https://github.com/edgexfoundry/docker-edgex-consul) to provide a public artifact that can be referenced by the blackbox testing's docker-compose.yml file to address race condition in container startup.  Related https://github.com/edgexfoundry/developer-scripts/pull/147.

**Comment by michaelestrin** [comment]
Tested on Linux system (special thank you to @AnthonyMBonafide). Was able to run the non-ARM64 non-security variant of the build by hand.  

Security version will require additional work (vault container is missing a volume mapping, environment variable to turn off security needs to be removed, consul needs to be updated to use the artifact from the new docker-edgex-consul pipeline, and https://github.com/edgexfoundry/developer-scripts/pull/147 needs to be merged).

@cloudxxx8 - please merge #291 and examine resulting non-security variant of tests before tomorrow's 6:00AM CST call.

cc: @tingyuz @jpwhitemn @tsconn23 @difince

**Comment by michaelestrin** [comment]
> @cloudxxx8 - please merge #291 and examine resulting non-security variant of tests before tomorrow's 6:00AM CST call.

After realizing I had write access, I merged the PR.  Waiting to see if tests actually run in Jenkins as a result.... https://jenkins.edgexfoundry.org/view/blackbox-testing/job/blackbox-testing-centos7-blackbox-4c-2g-master/597/console

**Comment by michaelestrin** [comment]
Follow up to this morning's call to discuss outstanding issues with Jenkins blackbox test builds.

To ensure the security version of the Jenkins builds are exercising the security features, I created an issue (https://github.com/edgexfoundry/blackbox-testing/issues/294) to parameterize the EDGEX_SECURITY_SECRET_STORE environment variable that's being set inside blackbox-testing's docker-compose.yml file.  I will work this issue with @AnthonyMBonafide to ensure it is merged into master, with @JPWKU on my team to force the build to run, and with @tingyuz to review the run's results.  I will send out a follow up email with those results once this action item is complete.  **_This needs to be completed before the Fuji release branches are cut._**

I created an issue (https://github.com/edgexfoundry/developer-scripts/issues/154) to add arm64 versions of the docker-compose files for the nightly nexus builds.  Although I will work this today, it does not need to be completed for the Fuji release.

I created an issue (https://github.com/edgexfoundry/blackbox-testing/issues/295) to modify the blackbox test scripts to use the appropriate developer-scripts' docker-compose file to bring up EdgeX.  I assigned it to @cloudxxx8 for your team to work post-Fuji release.

During our call, I thought the failed tests in the non-security arm64 build were specific to the applications-related tests and took an action item to reach out to Lenny.  On further review of the build output, the failures are all related to the rules engine.  I created an issue (https://github.com/edgexfoundry/blackbox-testing/issues/296) to address this.  **_Branching for Fuji should not be dependent upon completion of this issue though the work for this issue should be ported to the Fuji branch._**

**Comment by michaelestrin** [comment]
One additional item.



https://github.com/edgexfoundry/ci-management/pull/500​ was merged to modify the non-security Jenkins builds to add the EDGEX_SECURITY_SECRET_STORE=false environment variable required to turn off security.  Unfortunately, blackbox-testing's docker-compose doesn't take into account the Jenkins environment and this PR has no effect on the EdgeX containers.  We're solving this via https://github.com/edgexfoundry/blackbox-testing/issues/294.  To avoid future confusion around having this environment variable set in the Jenkins build (which does nothing), https://github.com/edgexfoundry/ci-management/pull/500​ should be reverted.  I will undertake this work.

**Comment by michaelestrin** [comment]
Quick update.

Security has been enabled in the secure blackbox test Jenkins builds.  Both are failing.  Troubleshooting is ongoing.  Believe issue is related to https://github.com/edgexfoundry/edgex-go/issues/1981 and https://github.com/edgexfoundry/edgex-go/pull/1945

**Comment by michaelestrin** [comment]
Additional manual testing of the security variation of the Jenkins blackbox testing builds found additional issues in both edgexfoundry/edgex-go (https://github.com/edgexfoundry/edgex-go/pull/1983) and edgexfoundry/blackbox-testing (https://github.com/edgexfoundry/blackbox-testing/pull/299) repos.

Working through the PR process for these now.  

Once these are merged, we'll need to manually kick off the nightly build for EdgeX to ensure the images pulled by the blackbox tests contain these latest code changes.  Once new images are available, we need to manually kick off the blackbox tests to ensure the changes fix the issues with the security variations of the Jenkins blackbox test builds.

Or we can ensure these get merged today and wait until tomorrow (after the nightly builds are run) to validate the changes.

**Comment by michaelestrin** [comment]
Processed PRs, merged, and managed to break the non-security build.  Duplicated the build on a Linux machine and had it run properly.  Added debug information to the blackbox test scripts and determined the build is broken for a reason we hadn't seen before. Core services are stopping because they can't parse secret store configuration.  Will continue investigation tomorrow.

**Comment by michaelestrin** [comment]
`17:03:39 SECURITY_IS_ON=false`

Unlike previous problem that broke the build, mongo container is running.

```
17:06:36 -------------------------- CORE DATA -----------------------
17:06:36 level=ERROR ts=2019-10-23T22:04:12.371434116Z app=edgex-core-data source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:12.394520059Z app=edgex-core-data source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 level=ERROR ts=2019-10-23T22:04:15.441725526Z app=edgex-core-data source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:15.480107948Z app=edgex-core-data source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 level=ERROR ts=2019-10-23T22:04:20.188618921Z app=edgex-core-data source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:20.236258019Z app=edgex-core-data source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 level=ERROR ts=2019-10-23T22:04:29.884488262Z app=edgex-core-data source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:29.911676152Z app=edgex-core-data source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 level=ERROR ts=2019-10-23T22:04:52.434884965Z app=edgex-core-data source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:52.483172712Z app=edgex-core-data source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 -------------------------------------------------------------
17:06:36 -------------------------- CORE METADATA -----------------------
17:06:36 level=ERROR ts=2019-10-23T22:04:40.587831503Z app=edgex-core-metadata source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:40.616769982Z app=edgex-core-metadata source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 level=ERROR ts=2019-10-23T22:04:44.466191073Z app=edgex-core-metadata source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:44.484983909Z app=edgex-core-metadata source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 level=ERROR ts=2019-10-23T22:04:48.065054039Z app=edgex-core-metadata source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:48.085005669Z app=edgex-core-metadata source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 level=ERROR ts=2019-10-23T22:04:52.198992859Z app=edgex-core-metadata source=logger.go:105 msg="logTarget cannot be blank, using stdout only"
17:06:36 level=ERROR ts=2019-10-23T22:04:52.245589754Z app=edgex-core-metadata source=secret.go:58 msg="unable to parse secret store configuration: unexpected end of JSON input"
17:06:36 ------------------------------------------------------------------
```

**Comment by michaelestrin** [comment]
Post merge of https://github.com/edgexfoundry/edgex-go/issues/1986, manually ran the nightly builds to update nexus images.  This fixed the issue with the non-security variation of the blackbox tests. 

Found issue with core-command attempting to access core-metadata's database.  Updated core-command to use core-metadata's credentials (each service now has its own Mongo credentials). Post merge of https://github.com/edgexfoundry/edgex-go/pull/1993, manually ran the nightly builds to update nexus images.  Non-secure build ran as expected.  Secure build failed with same errors as last night.  

Working theory is vault is not ready in time for vault-worker to do its work.  In https://github.com/edgexfoundry/blackbox-testing/pull/301, we added artificial delay between vault and vault-worker container starts, changed some init flag values, removed environment variable debugging, and added a dump of each container's logs to the build output.  Also found separate issue with rules-engine image version (duplicated in the developer-scripts repo and fixed via https://github.com/edgexfoundry/developer-scripts/issues/160).  Subsequently merged and both secure and insecure variants of the blackbox tests are running properly.

**Comment by michaelestrin** [comment]
Blackbox tests -- both secure and insecure versions -- are running properly for non-AMD64.  @tingyuz says as far as he knows we haven't run the AMD64 secure version successfully as there wasn't a compatible version of Kong.

https://jenkins.edgexfoundry.org/view/blackbox-testing/job/blackbox-testing-centos7-blackbox-4c-2g-master/611/allure/

https://jenkins.edgexfoundry.org/view/blackbox-testing/job/blackbox-testing-security-centos7-blackbox-4c-2g-master/325/allure/

**Comment by anonymouse64** [comment]
> Blackbox tests -- both secure and insecure versions -- are running properly for non-AMD64. @tingyuz says as far as he knows we haven't run the AMD64 secure version successfully as there wasn't a compatible version of Kong.

I belive you mean ARM64, not AMD64 here

**Comment by jamesrgregg** [comment]
> @tingyuz says as far as he knows we haven't run the AMD64 secure version successfully as there wasn't a compatible version of Kong.

@ernestojeda fixed the Kong image a few weeks ago and had it running just fine on ARM64 VM in Jenkins sandbox.  It's the officially supported Kong image.  
Please reference Issue #185

**Comment by michaelestrin** [comment]
> @ernestojeda fixed the Kong image a few weeks ago and had it running just fine on ARM64 VM in Jenkins sandbox. It's the officially supported Kong image.

Secure version of arm64 Jenkins blackbox tests is [currently failing](https://jenkins.edgexfoundry.org/view/blackbox-testing/job/blackbox-testing-security-ubuntu18.04-docker-arm64-4c-2g-master/294/consoleFull).  Unsecure version of arm64 (and both secure and unsecure non-arm64) Jenkins blackbox tests are running properly.

Services are unable to get vault token. Not sure why.

Are we sure we've been able to run the secure version of EdgeX on arm64?  This failure issue appears isolated to that context.

@tingyuz @ernestojeda @jamesrgregg @cloudxxx8

**Comment by michaelestrin** [comment]
@jpwhitemn see above.  I personally don't see the lack of resolution of the secure arm64 Jenkins blackbox test failure as needing to block the Fuji release.

**Comment by cloudxxx8** [comment]
We identified the error and opened https://github.com/edgexfoundry/blackbox-testing/issues/302 for arm64 security test

**Comment by michaelestrin** [comment]
Awesome @cloudxxx8.  Closing this issue.
